generate migration guide for Az 12.0.0 (#24877)
Azure/azure-powershell: documentation/migration-guides/Az.12.0.0.0-migration-guide.md

> # Migration Guide for Az 12.0.0.0

Reviewer: ```suggestion
# Migration Guide for Az 12.0.0
```

```
#### After
```powershell
error record will be disabled by default. New parameter 'EnableErrorRecordsPersistence' added to enable erro record.
```

## Az.Compute

### `Get-AzVmss`

- Cmdlet breaking-change will happen to all parameter sets
  - Starting in May 2024 the "Get-AzVmss" cmdlet will no longer allow an empty value for resource group name and virtual machine scale set name.
  - This change is expected to take effect from Az.Compute version: 8.0.0 and Az version: 12.0.0

#### Before
```powershell
Get-AzVmss -ResourceGroupName "" returned empty list
```
#### After
```powershell
Get-AzVmss -ResourceGroupName "" will return error for empty string validation in the parameter 
```

### `New-AzGalleryImageDefinition`

- Cmdlet breaking-change will happen to all parameter sets
  - Starting in May 2024 the "New-AzGalleryImage" cmdlet will deploy with the Trusted Launch configuration by default. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch
  - This change is expected to take effect from Az.Compute version: 8.0.0 and Az version: 12.0.0

#### Before
```powershell

> New-AzGalleryImageDefinition -ResourceGroupName $rgName -GalleryName $galleryName -Name $galleryImageDefinitionName -Location $location -Publisher $publisherName -Offer $offerName -Sku $skuName -OsState "Specialized" -OsType "Linux"                   defaulted to HyperVGeneration Gen1 and SecurityType Standard  in the service side 

Reviewer 2: I put some unnecesarry space after "Linux" in the excel sheet here, but maybe we remove it here?

Author: It's removed. Please check it.

```
#### After
```powershell

> Will default to HyperVGeneration Gen2 and SecurityType TrustedLaunchSupported by code in service side 

Reviewer 2: Same cmdlet let above will default to -

Author: updated. Please check it.

```

### `New-AzVM`

- Cmdlet breaking-change will happen to all parameter sets
  - Starting in May 2024 the "New-AzVM" cmdlet will deploy with the image 'Windows Server 2022 Azure Edition' by default. This will make migrating to Trusted Launch easier in the future. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch
  - This change is expected to take effect from Az.Compute version: 8.0.0 and Az version: 12.0.0

#### Before
```powershell

> $vm = New-AzVM -ResourceGroupName $rgname -Name $vmname -Credential $cred -SecurityType "Standard" -DomainNameLabel $domainNameLabel; Creates a VM with the image Windows 2016 Datacenter. 

Reviewer 3: ```suggestion
$vm = New-AzVM -ResourceGroupName $rgname -Name $vmname -Credential $cred -SecurityType "Standard" -DomainNameLabel $domainNameLabel; 
#Creates a VM with the image Windows 2016 Datacenter. 
```

```
#### After
```powershell

> $vm = New-AzVM -ResourceGroupName $rgname -Name $vmname -Credential $cred -SecurityType $securityTypeST -DomainNameLabel $domainNameLabel; Now generates with the Windows 2022 Azure Edition image.

Reviewer 3: ```suggestion
$vm = New-AzVM -ResourceGroupName $rgname -Name $vmname -Credential $cred -SecurityType $securityTypeST -DomainNameLabel $domainNameLabel; 
#Now generates with the Windows 2022 Azure Edition image.
```

```

### `New-AzVmss`

- Cmdlet breaking-change will happen to all parameter sets
  - Starting in May 2024 the "New-AzVmss" cmdlet will deploy with the image 'Windows Server 2022 Azure Edition' by default. This will make migrating to Trusted Launch easier in the future. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch
  - This change is expected to take effect from Az.Compute version: 8.0.0 and Az version: 12.0.0

#### Before
```powershell

> $vmss = New-AzVmss -ResourceGroupName $rgname -Credential $cred -VMScaleSetName $vmssName -SecurityType "Standard" -DomainNameLabel $domainNameLabel1; would create a VMSS with a Windows 2016 Datacenter image. 

Reviewer 3: ```suggestion
$vmss = New-AzVmss -ResourceGroupName $rgname -Credential $cred -VMScaleSetName $vmssName -SecurityType "Standard" -DomainNameLabel $domainNameLabel1; 
#Creates a VMSS with a Windows 2016 Datacenter image. 
```

```
#### After
```powershell

> $vmss = New-AzVmss -ResourceGroupName $rgname -Credential $cred -VMScaleSetName $vmssName -SecurityType "Standard" -DomainNameLabel $domainNameLabel1; Now it will generate with a Windows 2022 Azure Edition image.

Reviewer 3: ```suggestion
$vmss = New-AzVmss -ResourceGroupName $rgname -Credential $cred -VMScaleSetName $vmssName -SecurityType "Standard" -DomainNameLabel $domainNameLabel1; 
#Now it will generate with a Windows 2022 Azure Edition image.
```

  - This change is expected to take effect from Az.Storage version: 7.0.0 and Az version: 12.0.0

#### Before
```powershell
PS C:\WINDOWS\system32> $queue = Get-AzStorageQueue -Context $ctx-Name myqueue
PS C:\WINDOWS\system32> $queue | fl

CloudQueue              : Microsoft.Azure.Storage.Queue.CloudQueue
Uri                     : https://mystorageaccount.queue.core.windows.net/myqueue
ApproximateMessageCount : 0
EncodeMessage           : True
QueueClient             : Azure.Storage.Queues.QueueClient
QueueProperties         : Azure.Storage.Queues.Models.QueueProperties
Context                 : Microsoft.WindowsAzure.Commands.Storage.AzureStorageContext
Name                    : myqueue

PS C:\WINDOWS\system32> $queueMessage = [Microsoft.Azure.Storage.Queue.CloudQueueMessage]::new("This is message 1")
PS C:\WINDOWS\system32> $queue.CloudQueue.AddMessageAsync($queueMessage)
```
#### After
```powershell
PS C:\WINDOWS\system32> $queue = Get-AzStorageQueue -Context $ctx-Name myqueue
PS C:\WINDOWS\system32> $queue | fl

QueueClient             : Azure.Storage.Queues.QueueClient

> Uri                     : https://mysorageaccount.queue.core.windows.net/myqueue

Reviewer 4: Just noticed I had a typo here -- It should be mystorageaccount instead of mysorageaccount. 
Is it still possible for us to fix it?

```powershell
error record will be disabled by default. New parameter 'EnableErrorRecordsPersistence' added to enable erro record.
```

### `Update-AzConfig`

- Cmdlet breaking-change will happen to all parameter sets
  - Parameter `DisableErrorRecordsPersistence` will be deprecated, a new parameter `EnableErrorRecordsPersistence` will be added instead. Writing error records to file system will become opt-in instead of opt-out. This change will happen around May 2024
  - This change is expected to take effect from Az.Accounts version: 2.X and Az version: 12.0.0

#### Before
```powershell
Update-AzConfig -DisableErrorRecordsPersistence 
```
#### After
```powershell
error record will be disabled by default. New parameter 'EnableErrorRecordsPersistence' added to enable erro record.
```

## Az.Compute

### `Get-AzVmss`

- Cmdlet breaking-change will happen to all parameter sets
  - Starting in May 2024 the "Get-AzVmss" cmdlet will no longer allow an empty value for resource group name and virtual machine scale set name.
  - This change is expected to take effect from Az.Compute version: 8.0.0 and Az version: 12.0.0

#### Before
```powershell
Get-AzVmss -ResourceGroupName ""

> # returned empty list

Reviewer 2: ```suggestion
# Returned an empty list.
```

```
#### After
```powershell
Get-AzVmss -ResourceGroupName ""

> # will return error for empty string validation in the parameter 

Reviewer 2: ```suggestion
# Will return an error from empty string validation in the parameter.
```

```

### `New-AzGalleryImageDefinition`

- Cmdlet breaking-change will happen to all parameter sets

>   - Starting in May 2024 the "New-AzGalleryImage" cmdlet will deploy with the Trusted Launch configuration by default. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch

Reviewer 2: ```suggestion
  - Starting in May 2024 the "New-AzGalleryImage" cmdlet will deploy with the Trusted Launch configuration and Gen2 Hyper V Generation by default. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch
```

Reviewer 5: ```suggestion
  - Starting in May 2024 the "New-AzGalleryImage" cmdlet will deploy with the Trusted Launch configuration and Gen2 Hyper V Generation by default. To know more about Trusted Launch, please visit [https://learn.microsoft.com/azure/virtual-machines/trusted-launch](/azure/virtual-machines/trusted-launch)
```

Reviewer 5: This URL causes a hard-coded-locale and docs-link-absolute build warning in the docs repo.

  - This change is expected to take effect from Az.Compute version: 8.0.0 and Az version: 12.0.0

#### Before
```powershell
New-AzGalleryImageDefinition -ResourceGroupName $rgName -GalleryName $galleryName -Name $galleryImageDefinitionName -Location $location -Publisher $publisherName -Offer $offerName -Sku $skuName -OsState "Specialized" -OsType "Linux"

> # defaulted to HyperVGeneration Gen1 and SecurityType Standard  in the service side 

Reviewer 2: ```suggestion
# Defaulted to HyperVGeneration: Gen1 and SecurityType: Standard in the service side .
```

```
#### After
```powershell

> # Will default to HyperVGeneration Gen2 and SecurityType TrustedLaunchSupported by code in service side 

Reviewer 2: ```suggestion
New-AzGalleryImageDefinition -ResourceGroupName $rgName -GalleryName $galleryName -Name $galleryImageDefinitionName -Location $location -Publisher $publisherName -Offer $offerName -Sku $skuName -OsState "Specialized" -OsType "Linux"
# Defaults to HyperVGeneration: Gen2 and SecurityType: TrustedLaunchSupported at the PowerShell level.
```

```

### `New-AzVM`

- Cmdlet breaking-change will happen to all parameter sets
  - Starting in May 2024 the "New-AzVM" cmdlet will deploy with the image 'Windows Server 2022 Azure Edition' by default. This will make migrating to Trusted Launch easier in the future. To know more about Trusted Launch, please visit https://docs.microsoft.com/en-us/azure/virtual-machines/trusted-launch
  - This change is expected to take effect from Az.Compute version: 8.0.0 and Az version: 12.0.0

#### Before
```powershell
$vm = New-AzVM -ResourceGroupName $rgname -Name $vmname -Credential $cred -SecurityType "Standard" -DomainNameLabel $domainNameLabel
# Creates a VM with the image Windows 2016 Datacenter. 
```
#### After
```powershell

> $vm = New-AzVM -ResourceGroupName $rgname -Name $vmname -Credential $cred -SecurityType $securityTypeST -DomainNameLabel $domainNameLabel

Author: done

  - `-InboundIpRule`
    - The parameter : 'InboundIpRule' is changing.
    The type of the parameter is changing from 'Microsoft.Azure.Commands.EventGrid.Models.PSInboundIpRule[]' to 'IInboundIPRule[]'.
    - This change is expected to take effect from Az.EventGrid version: 2.0.0 and Az version: 12.0.0
  - `-PrivateEndpointConnection`
    - This parameter will be deprecated.
    - This change is expected to take effect from Az.EventGrid version: 2.0.0 and Az version: 12.0.0

#### Before
```powershell
New-AzEventGridPartnerNamespace -ResourceGroupName MyResourceGroupName -Name PartnerNamespace1 -Location westus2 -PartnerRegistrationFullyQualifiedId 23e0092b-f336-4833-9ab3-9353a15650fc
```
#### After
```powershell
New-AzEventGridPartnerNamespace -Name azps-partnernamespace -ResourceGroupName azps_test_group_eventgrid -Location eastus -PartnerTopicRoutingMode ChannelNameHeader -PartnerRegistrationFullyQualifiedId "/subscriptions/{subId}/resourceGroups/azps_test_group_eventgrid/providers/Microsoft.EventGrid/partnerRegistrations/azps-registration"
```

### `New-AzEventGridPartnerRegistration`

- Cmdlet breaking-change will happen to all parameter sets

>   - Added new required parameter: Location <String>

Reviewer 5: ```suggestion
  - Added new required parameter: Location `<String>`
```

Reviewer 5: This line causes a `disallowed-html-tag` docs build warning.

  - This change is expected to take effect from Az.Storage version: 7.0.0 and Az version: 12.0.0

#### Before
```powershell
PS C:\WINDOWS\system32> $queue = Get-AzStorageQueue -Context $ctx-Name myqueue
PS C:\WINDOWS\system32> $queue | fl

CloudQueue              : Microsoft.Azure.Storage.Queue.CloudQueue
Uri                     : https://mystorageaccount.queue.core.windows.net/myqueue
ApproximateMessageCount : 0
EncodeMessage           : True
QueueClient             : Azure.Storage.Queues.QueueClient
QueueProperties         : Azure.Storage.Queues.Models.QueueProperties
Context                 : Microsoft.WindowsAzure.Commands.Storage.AzureStorageContext
Name                    : myqueue

PS C:\WINDOWS\system32> $queueMessage = [Microsoft.Azure.Storage.Queue.CloudQueueMessage]::new("This is message 1")
PS C:\WINDOWS\system32> $queue.CloudQueue.AddMessageAsync($queueMessage)
```
#### After
```powershell
PS C:\WINDOWS\system32> $queue = Get-AzStorageQueue -Context $ctx-Name myqueue
PS C:\WINDOWS\system32> $queue | fl

QueueClient             : Azure.Storage.Queues.QueueClient

> Uri                     : https://mysorageaccount.queue.core.windows.net/myqueue

Reviewer 4: ```suggestion
Uri                     : https://mystorageaccount.queue.core.windows.net/myqueue
```
